src/Common/src/System/Security/Cryptography/AsnReader.cs

```diff
@@ -0,0 +1,2773 @@
+            }
+
+            totalBytesRead = headerLength + length.Value;
+            return builder.ToString();
+        }
+
+        public string ReadObjectIdentifierAsString() =>
+            ReadObjectIdentifierAsString(Asn1Tag.ObjectIdentifier);
+        
+        public string ReadObjectIdentifierAsString(Asn1Tag expectedTag)
+        {
+            string oidValue = ReadObjectIdentifierAsString(expectedTag, out int bytesRead);
+
+            _data = _data.Slice(bytesRead);
+
+            return oidValue;
+        }
+
+        public Oid ReadObjectIdentifier(bool skipFriendlyName = false) =>
+            ReadObjectIdentifier(Asn1Tag.ObjectIdentifier, skipFriendlyName);
+
+        public Oid ReadObjectIdentifier(Asn1Tag expectedTag, bool skipFriendlyName=false)
+        {
+            string oidValue = ReadObjectIdentifierAsString(expectedTag, out int bytesRead);
+            Oid oid = skipFriendlyName ? new Oid(oidValue, oidValue) : new Oid(oidValue);
+
+            // Don't slice until the return object has been created.
+            _data = _data.Slice(bytesRead);
+
+            return oid;
+        }
+
+        private bool TryCopyCharacterStringBytes(
+            Asn1Tag expectedTag,
+            UniversalTagNumber universalTagNumber,
+            Span<byte> destination,
+            bool write,
+            out int bytesRead,
+            out int bytesWritten)
+        {
+            if (TryGetOctetStringBytes(
+                expectedTag,
+                out ReadOnlyMemory<byte> contents,
+                out int headerLength,
+                universalTagNumber))
+            {
+                bytesWritten = contents.Length;
+
+                if (write)
+                {
+                    if (destination.Length < bytesWritten)
+                    {
+                        bytesWritten = 0;
+                        bytesRead = 0;
+                        return false;
+                    }
+
+                    contents.Span.CopyTo(destination);
+                }
+
+                bytesRead = headerLength + bytesWritten;
+                return true;
+            }
+
+            Asn1Tag tag = ReadTagAndLength(out int ? length, out headerLength);
+
+            bool copied = TryCopyConstructedOctetStringValue(
+                Slice(_data, headerLength, length),
+                destination,
+                write,
+                _ruleSet,
+                length == null,
+                out int contentBytesRead,
+                out bytesWritten);
+
+            bytesRead = headerLength + contentBytesRead;
+            return copied;
+        }
+
+        private static unsafe string GetCharacterString(
+            ReadOnlyMemory<byte> source,
+            Text.Encoding encoding)
+        {
+            fixed (byte* bytePtr = &source.Span.DangerousGetPinnableReference())
+            {
+                try
+                {
+                    return encoding.GetString(bytePtr, source.Length);
```

Comment:
For this and the method below, would it make sense to use the new span-based parsers?

Reply:
The Utf8Parser class turns UTF-8 text (in a ReadOnlySpan<byte>) to fixed size structures (integral types, bool, DateTime/Offset, Guid, TimeSpan, fractional types).

These particular methods are for "textual data", like a certificate subject name('s fields); so the System.Text.Encoding classes are the most appropriate (with their Span-based goodness in netcoreapp-only)